scylla/src/transport/query_result.rs

```diff
@@ -0,0 +1,210 @@
+use scylla_cql::types::deserialize::row::{DeserializeRow, TypedRowIterator};
+
+use super::legacy_query_result::Legacy08QueryResult;
+
+/// Raw results of a single query.
+///
+/// More comprehensive description TODO
+#[derive(Default, Debug)]
+pub struct QueryResult {
+    raw_rows: Option<RawRows>,
+    tracing_id: Option<Uuid>,
+    warnings: Vec<String>,
+}
+
+impl QueryResult {
+    /// Returns the number of received rows, or `None` if the response wasn't of Rows type.
+    pub fn rows_num(&self) -> Option<usize> {
+        Some(self.raw_rows.as_ref()?.rows_count())
+    }
+
+    /// Returns a bool indicating the current response is of Rows type.
+    pub fn is_rows(&self) -> bool {
+        self.raw_rows.is_some()
+    }
+
+    /// Returns the received rows when present.
+    ///
+    /// Returns an error if the original query didn't return a Rows response (e.g. it was an `INSERT`),
+    /// or the response is of incorrect type.
+    pub fn rows<'s, R: DeserializeRow<'s>>(&'s self) -> Result<TypedRowIterator<'s, R>, RowsError> {
+        Ok(self
+            .raw_rows
+            .as_ref()
+            .ok_or(RowsError::NotRowsResponse)?
+            .rows_iter()?)
+    }
+
+    /// Returns the received rows when present, or None.
+    ///
+    /// Returns an error if the rows are present but are of incorrect type.
+    pub fn maybe_rows<'s, R: DeserializeRow<'s>>(
+        &'s self,
+    ) -> Result<Option<TypedRowIterator<'s, R>>, ParseError> {
+        match &self.raw_rows {
+            Some(rows) => Ok(Some(rows.rows_iter()?)),
+            None => Ok(None),
+        }
+    }
+
+    /// Returns `Ok` for a result of a query that shouldn't contain any rows.\
+    /// Will return `Ok` for `INSERT` result, but a `SELECT` result, even an empty one, will cause an error.\
+    /// Opposite of [`rows()`](QueryResult::rows).
+    pub fn result_not_rows(&self) -> Result<(), ResultNotRowsError> {
+        match &self.raw_rows {
+            Some(_) => Err(ResultNotRowsError),
+            None => Ok(()),
+        }
+    }
+
+    /// Returns first row from the received rows.\
+    /// When the first row is not available, returns an error.
+    pub fn first_row<'s, R: DeserializeRow<'s>>(&'s self) -> Result<R, FirstRowError> {
+        match self.maybe_first_row::<R>() {
+            Ok(Some(row)) => Ok(row),
+            Ok(None) => Err(FirstRowError::RowsEmpty),
+            Err(RowsError::NotRowsResponse) => Err(FirstRowError::NotRowsResponse),
+            Err(RowsError::TypeCheckFailed(err)) => Err(FirstRowError::TypeCheckFailed(err)),
+        }
+    }
+
+    /// Returns `Option<R>` containing the first of a result.\
+    /// Fails when the query isn't of a type that could return rows, same as [`rows()`](QueryResult::rows).
+    pub fn maybe_first_row<'s, R: DeserializeRow<'s>>(&'s self) -> Result<Option<R>, RowsError> {
+        Ok(self.rows::<R>()?.next().transpose()?)
+    }
```

Comment:
Shouldn't `FirstRowError` be returned from `maybe_first_row` function?

Reply:
`FirstRowError` - no, it is too broad, it can represent some failures which cannot happen with `maybe_first_row`.

Maybe you meant to say that `maybe_first_row` should return an error if the response was not of Rows type, like the comment suggests? In that case I agree and will fix that.

Reply:
I will add a `MaybeFirstRowError` struct.